components/core/src/clp/string_utils/string_utils.cpp

```diff
@@ -182,114 +182,157 @@ bool wildcard_match_unsafe(string_view tame, string_view wild, bool case_sensiti
-                // ensures there are no dangling escape characters
-                w = *wild_current;
+                // Safe without a bounds check
+                ++wild_it;
+                w = *wild_it;
             }
 
             // Handle a mismatch
-            t = *tame_current;
-            if (!((false == is_escaped && '?' == w) || t == w)) {
-                if (nullptr == wild_bookmark) {
-                    // No bookmark to return to
+            if (w != *tame_it) {
+                // Reset to bookmarks
+                tame_it = tame_bookmark_it + 1;
+                if (tame_end_it == tame_it) {
                     return false;
                 }
-
-                wild_current = wild_bookmark;
-                tame_current = tame_bookmark + 1;
+                wild_it = wild_bookmark_it;
                 if (false
-                    == advance_tame_to_next_match(
-                            tame_current,
-                            tame_bookmark,
-                            tame_end,
-                            wild_current
-                    ))
+                    == advance_tame_to_next_match(tame_end_it, tame_it, tame_bookmark_it, wild_it))
                 {
                     return false;
                 }
             }
         }
 
-        ++tame_current;
-        ++wild_current;
+        ++tame_it;
+        ++wild_it;
 
-        // Handle reaching the end of tame or wild
-        if (tame_end == tame_current) {
-            return (wild_end == wild_current
-                    || ('*' == *wild_current && (wild_current + 1) == wild_end));
-        } else {
-            if (wild_end == wild_current) {
-                if (nullptr == wild_bookmark) {
-                    // No bookmark to return to
-                    return false;
-                } else {
-                    wild_current = wild_bookmark;
-                    tame_current = tame_bookmark + 1;
-                    if (false
-                        == advance_tame_to_next_match(
-                                tame_current,
-                                tame_bookmark,
-                                tame_end,
-                                wild_current
-                        ))
-                    {
-                        return false;
-                    }
-                }
+        // Handle boundary conditions
+        if (tame_end_it == tame_it) {
+            return (wild_end_it == wild_it) || (wild_end_it == wild_it + 1 && '*' == *wild_it);
+        } else if (wild_end_it == wild_it) {
+            if (tame_end_it == tame_it) {
+                return true;
+            }
+
+            // Reset to bookmarks
+            tame_it = tame_bookmark_it + 1;
+            wild_it = wild_bookmark_it;
+            if (false
+                == advance_tame_to_next_match(tame_end_it, tame_it, tame_bookmark_it, wild_it))
+            {
+                return false;
```

Comment:
Correct me if I'm wrong: if this branch is triggered, wild has already reached the end without consuming the entire tame. We should be handling the last group of tame after the last `*`. In this case, we only need to match the last n characters (determined by `wild_it - wild_bookmark_it`, and properly counting escape chars in between) in tame right? For example, if tame is "aaaaaaaa" and wild is "*a", we don't have to advance tame to match every single "a" but jump to match the last one

Reply:
You're right. Are you suggesting, in this case, that we should iterate backwards from the end of `tame` to see if it matches the last group in `wild`?

Reply:
Actually, iterating backwards is non-trivial because of escaped characters. If we see a `?` in `wild`, we have to check the character before it to know if it's an escape character. But even if it is, we don't know if it's escaping the `?` or it's preceded by an escape itself. So it's easier to always iterate forwards.